delft/sequenceLabelling/data_generator.py

```diff
@@ -52,26 +54,21 @@ def __getitem__(self, index):
         if self.preprocessor.return_casing:
             batch_x, batch_c, batch_a, batch_l, batch_y = self.__data_generation(index)
             return [batch_x, batch_c, batch_a, batch_l], batch_y
-        else:  
+        else:
             batch_x, batch_c, batch_l, batch_y = self.__data_generation(index)
             return [batch_x, batch_c, batch_l], batch_y
 
-    def shuffle_pair(self, a, b):
-        # generate permutation index array
-        permutation = np.random.permutation(a.shape[0])
-        # shuffle the two arrays
-        return a[permutation], b[permutation]
-
     def on_epoch_end(self):
+        # If we are predicting, we don't need to shuffle
+        if self.y is None:
+            return
+
         # shuffle dataset at each epoch
-        if self.shuffle == True:
-            if self.y is None:
-                np.random.shuffle(self.x)
-            else:      
-                self.shuffle_pair(self.x,self.y)
+        if self.shuffle:
+            self.x, self.y = shuffle_pair_with_view(self.x, self.y)
```

Comment:
I don't know enough about how numpy views work internally. i.e. whether a numpy view of a view references the view, in which case this would create an ever longer chain of views; or whether it will only reference the original data.

Because I am not sure how it works, I would maybe recommend saving the original values to say `orginal_x` and `original_y`.

And change this line to:

```python
self.x, self.y = shuffle_pair_with_view(self.original_x, self.original_y)
```

(Otherwise it might be worth researching the answer, unless you already know it)

Reply:
Thanks. I will have a look.